lib/service/cfg.go

```diff
@@ -277,20 +277,25 @@ func (cfg *Config) ApplyToken(token string) (bool, error) {
 }
 
 // ApplyCAPins assigns the given CA pin(s), filtering out empty pins.
+// If a pin is specified as a path to a file, that file must not be empty.
 func (cfg *Config) ApplyCAPins(caPins []string) error {
-	if len(caPins) == 0 {
+	if len(caPins) == 0 || (len(caPins) == 1 && caPins[0] == "") {
 		return nil
 	}
 	filteredPins := make([]string, 0, len(caPins))
-	for _, pin := range caPins {
-		if pin != "" {
-			var err error
-			pin, err = utils.TryReadValueAsFile(pin)
-			if err != nil {
-				return trace.Wrap(err)
-			}
-			filteredPins = append(filteredPins, pin)
+	for _, pinOrPath := range caPins {
+		if pinOrPath == "" {
+			continue
```

Comment:
Should we error here? We already cleared the "len = 1 && empty" case above, so we know this is a strange looking config (that will likely fail further down the line)?

You could just let it go through TryReadValueAsFile as well, and let it error on line 295.

Reply:
The idea was that empty ca_pin values are fine, but empty ca_pin files are probably not expected and should produce an error. It's a little strange that prior behavior would allow one and only one empty ca_pin but not allow it if there are more than one. I suppose if we kept this then we could remove the "if len == 1" check above though.

This won't break any existing teleport configs if we allow empty values in the yaml. I don't have a strong preference though, so we can error here if you prefer that.

Reply:
> Let's make sure ApplyCAPins keeps the behavior of the old if len > 1, preferably erroring if any pins are empty in that scenario

My bad, I totally agreed with you above and then submitted code which doesn't error for empty pins. During a 1x1 with Roman, we discussed this and decided on allowing empty ca_pin values if it's directly in the yaml and error only for an empty ca_pin file. We agreed on that because it's more consistent blank ca_pin yaml behavior and wouldn't break any existing configs.